logging/logging.go

```diff
@@ -0,0 +1,134 @@
+package logging
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+)
+
+type BackupSummary struct {
+	MessageType         string  `json:"message_type"`
+	FilesNew            int     `json:"files_new"`
+	FilesChanged        int     `json:"files_changed"`
+	FilesUnmodified     int     `json:"files_unmodified"`
+	DirsNew             int     `json:"dirs_new"`
+	DirsChanged         int     `json:"dirs_changed"`
+	DirsUnmodified      int     `json:"dirs_unmodified"`
+	DataBlobs           int     `json:"data_blobs"`
+	TreeBlobs           int     `json:"tree_blobs"`
+	DataAdded           int64   `json:"data_added"`
+	TotalFilesProcessed int     `json:"total_files_processed"`
+	TotalBytesProcessed int     `json:"total_bytes_processed"`
+	TotalDuration       float64 `json:"total_duration"`
+	SnapshotID          string  `json:"snapshot_id"`
+}
+
+type BackupEnvelope struct {
+	MessageType string `json:"message_type,omitempty"`
+	BackupStatus
+	BackupSummary
+	BackupError
+}
+
+type BackupStatus struct {
+	PercentDone  float64  `json:"percent_done"`
+	TotalFiles   int      `json:"total_files"`
+	FilesDone    int      `json:"files_done"`
+	TotalBytes   int      `json:"total_bytes"`
+	BytesDone    int      `json:"bytes_done"`
+	CurrentFiles []string `json:"current_files"`
+	ErrorCount   int      `json:"error_count"`
+}
+
+type BackupOutputParser struct {
+	Log         logr.Logger
+	errorCount  int
+	lineCounter int
+	Summaryfunc func(summary BackupSummary, errorCount int, folder string, startTimestamp, endTimestamp int64)
+	Folder      string
+}
+
+type BackupError struct {
+	Error struct {
+		Op   string `json:"Op"`
+		Path string `json:"Path"`
+		Err  int    `json:"Err"`
+	} `json:"error"`
+	During string `json:"during"`
+	Item   string `json:"item"`
+}
+
+// LineParser takes a single string from an output and passes it the the
+// concrete implementation
+type LineParser interface {
+	Parse(s string) error
+}
+
+// OutputWrapper will split the output into lines.
+type OutputWrapper struct {
+	Parser LineParser
+}
+
+func (s *OutputWrapper) Write(p []byte) (n int, err error) {
+
+	scanner := bufio.NewScanner(bytes.NewReader(p))
+
+	for scanner.Scan() {
+		err := s.Parser.Parse(scanner.Text())
+		if err != nil {
+			return len(p), err
+		}
+	}
+
+	return len(p), nil
+}
+
+type LogOutParser struct {
+	Log logr.Logger
+}
+
+func (l *LogOutParser) Parse(s string) error {
+	l.Log.Info(s)
+	return nil
+}
+
+type LogErrParser struct {
+	Log logr.Logger
+}
+
+func (l *LogErrParser) Parse(s string) error {
```

Comment:
Parser seems to be a miss nomer. This is not parsing anything.

Reply:
I would get rid of the parser and just go with the wrapper. In the most simplest form, this can be just like https://github.com/corvus-ch/logr/blob/master/writer_adapter/writer.go. For the case with the JSON parsing, there is no need for this abstraction. Just handle it with a simple wrapper and a JSON parsing wrapper.

Reply:
I changed it to your proposed pattern. Makes it much easier to handle.

Reply:
It definitely is way simpler on the usage side now. 👏🏼